HW2/P5/physics.pyx

```diff
@@ -69,17 +86,96 @@ cdef void sub_update(FLOAT[:, ::1] XY,
     ############################################################
     # SUBPROBLEM 2: use the grid values to reduce the number of other
     # objects to check for collisions.
-    for j in range(i + 1, count):
-        XY2 = &(XY[j, 0])
-        V2 = &(V[j, 0])
-        if overlapping(XY1, XY2, R):
-            # SUBPROBLEM 4: Add locking
-            if not moving_apart(XY1, V1, XY2, V2):
-                collide(XY1, V1, XY2, V2)
-
-            # give a slight impulse to help separate them
-            for dim in range(2):
-                V2[dim] += eps * (XY2[dim] - XY1[dim])
+
+    # Instead of checking all i+1 remaining higher-ranked balls against
+    # the ith ball, locate the ith ball on the Grid (my_gloc), and then check
+    # neighboring grid spaces (up to two in each direction - a 5x5 grid
+    # centered at my_gloc) for balls -- this replaces the O(n) inner loop
+    # with a constant number (max. 24) of comparisons, thus making the
+    # overall collision detection algorithm O(n)
+    
+    gridsize = <int>((1.0 / grid_spacing) + 1)
+    my_gloc_x = <int>(XY[i, 0] / grid_spacing)
+    my_gloc_y = <int>(XY[i, 1] / grid_spacing)
+
+    # check neighboring grid cells, except cells out of boundaries,
+    # own cells, or empty cells (empty cells are denoted by -1 in Grid,
+    # or a very large number as an unsigned integer >> count)
+
+    left = my_gloc_x - 2
+    if left<0:
+        left = 0
+
+    right = my_gloc_x + 2
+    if right>=gridsize:
+        right = gridsize - 1 # inclusive
+    
+    bottom = my_gloc_y - 2
+    if bottom<0:
+        bottom = 0
+    
+    top = my_gloc_y + 2
+    if top>=gridsize:
+        top = gridsize - 1 # inclusive
+
+    for j in range(left, right+1):
+        for k in range(bottom, top+1):
```

Comment:
I like how you calculate the bounds before the loop, instead of doing all of the math in one or two lines. It's a little more verbose but clearer what's going on.